### ISSUE TITLE
NPSP Contact Merge May Not Honor Values Selected from Non-“Master” Record

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#6435]
> This issue was introduced in NPSP 3.198.

### Issue:

There are at least 3 scenarios where Contact Merge is not retaining the selected field value of a Contact in our Contact Merge feature:

1. Customer clicks the _Select All_ link above the non-“master” record and then selects a field value from another record. 
2. Customer manually selects a different “master” record by selecting the Id value in another column’s _Master Record Id_ row and then selects a different field value from that same Contact. 
3. Customer keeps the default “master” record selected and then selects a different field value from another column.

### Root Cause
**TL/DR:** We are not updating the selectedValue variable for the appropriate FieldRow when it changes. It is retaining its original Id value depending upon the scenario above.
*
**More Info**
When we call the mergeRecords() method in CON_ContactMerge_CTRL,  we set the master variable’s Id to the to the row where the master row’s showRadio value = true. We then loop through the rest of the FieldRow values and if the following condition is met, we set the master variable’s field values to the row’s selectedValue.
```
row.showRadio && 
!row.fieldName.equals(MASTER_KEY) && 
row.selectedValue != master.id
```

In the scenarios I’ve outlined, this condition is never met because the row’s selectedValue is retaining the master Id’s value. This results in an update to the master Id record which does not include a new value. 

The selectedValue variable for a FieldRow is set within the loadCandidates() method which gets called on initial load of the Contact Merge page which displays the Contacts you’re merging. It also gets set if you click the Select All link for a Contact  and the selectDefaultRecord() method. 

Ultimately, this selectedValue is not being reset for a selected FieldRow when it is not the “master” record.

*Sample Log Lines*
We have 2 Contacts. One where the Nickname is Jessie (Id = 0036300000nxO2HAAU) and one where it’s Jess (Id = 0036300000nxWsLAAU). When we search for these Contacts and select them, the Contact Merge page calls the loadMergeCandidate() method and sets the default “master” record to Jess since her record was most recently modified.
The fieldRow for the master row and the Nickname row at this point:
```
13:46:59:881 USER_DEBUG [708]|DEBUG|fieldrow:FieldRow:[fieldLabel=Master Record, fieldName=$MASTER$, selectedValue=0036300000nxWsLAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=null], Cell:[objId=0036300000nxWsLAAU, value=null])]
13:46:59:884 USER_DEBUG [708]|DEBUG|fieldrow:FieldRow:[fieldLabel=Nickname, fieldName=nickname__c, selectedValue=0036300000nxWsLAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=Jessie], Cell:[objId=0036300000nxWsLAAU, value=Jess])]
```
The selectedValue of both rows is the Jess record (0036300000nxWsLAAU), as expected.

When I click _Select All_ for the Jessie record, the selectDefaultRecord() method is called and we have the following values in our FieldRows:
```
13:48:33:057 USER_DEBUG [485]|DEBUG|*** in selectDefaultRecord. Record Id is: 0036300000nxO2HAAU
13:50:45:089 USER_DEBUG [495]|DEBUG|*** end of selectDefaultRecord. FieldRows = FieldRow:[fieldLabel=Master Record, fieldName=$MASTER$, selectedValue=0036300000nxO2HAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=null], Cell:[objId=0036300000nxWsLAAU, value=null])]
13:50:45:092 USER_DEBUG [495]|DEBUG|*** end of selectDefaultRecord. FieldRows = FieldRow:[fieldLabel=Nickname, fieldName=nickname__c, selectedValue=0036300000nxO2HAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=Jessie], Cell:[objId=0036300000nxWsLAAU, value=Jess])]
```
The selectedValue of both rows is the Jessie Id (0036300000nxWsLAAU), as expected.

If I then select the Nickname value from the Jess record and hit Merge, I see the following values in the mergeCandidates() method:
```
13:52:29:045 USER_DEBUG [957]|DEBUG|*** fieldRows = FieldRow:[fieldLabel=Master Record, fieldName=$MASTER$, selectedValue=0036300000nxO2HAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=null], Cell:[objId=0036300000nxWsLAAU, value=null])]
13:52:29:047 USER_DEBUG [957]|DEBUG|*** fieldRows = FieldRow:[fieldLabel=Nickname, fieldName=nickname__c, selectedValue=0036300000nxO2HAAU, showRadio=true, styleClass=null, values=(Cell:[objId=0036300000nxO2HAAU, value=Jessie], Cell:[objId=0036300000nxWsLAAU, value=Jess])]
```
The selectedValue for the Nickname__c field is still retaining the Jessie record, despite having selected the Jess value.

### Steps to Repeat
*Setup*:

* Create a custom text field on Contact called Nickname

*Test 1: Use Select All option for Non-“Master” Contact and select field value from other Contact record.*

*Expected Behavior:* the merged Contact will have “Jess” as the Nickname value.
*Actual Behavior*: the merged Contact still has “Jessie” as the Nickname value.

1. Create a Contact (First Name = Test; Last Name = Contact; Nickname = Jessie)
2. Create a Contact (First Name = Test; Last Name = Contact; Nickname = Jess)
3. Go to the Contact Merge Tab and click the Search Contacts button. Search for “Test Contact”. 
4. Select both records in the found set and hit Next.
5. The record selected within the Master Record row should be the 2nd one you Created since it defaults to the record with the most recent modified date. Its Nickname value of “Jess” should also be selected.
6. Use the “Select All” option above the non-“Master” record. This will reload the page with that first Contact’s values selected, including that Nickname value of “Jessie”.
7. Select the “Jess” Nickname value from the non-“Master” Contact.
8. Click the Merge button and hit Merge in the warning modal.
9. Confirm the merged Contact has a value of “Jessie” instead of “Jess”.

***
*Test 2: Keep the default “Master” record selected and choose a field value from the other Contact.*

*Expected Behavior:* the merged Contact will have “Jessie” as the Nickname value.
*Actual Behavior*: the merged Contact still has “Jess” as the Nickname value.

1. After the first test, you should only have 1 Contact where the Nickname is “Jessie”. Create a 2nd Contact (First Name = Test; Last Name = Contact; Nickname = Jess).
2. Go to the Contact Merge Tab and click the Search Contacts button. Search for “Test Contact”. 
3. Select both records in the found set and hit Next.
4. The record selected within the Master Record row should be the 2nd one you Created since it defaults to the record with the most recent modified date. Its Nickname value of “Jess” should also be selected.
5. Select the Nickname value of “Jessie” from the non-“Master” Contact.
6. Click the Merge button and hit Merge in the warning modal.
7. Confirm the merged Contact has a value of “Jess” instead of “Jessie”.

***

*Test 3: Select the Master Record Radio Button for the non-“Master” record and select the field value for Nickname from that record as well.* 

*Expected Behavior:* the merged Contact will have “Jess” as the Nickname value.
*Actual Behavior*: the merged Contact will have “Jessica” as the Nickname value.

1. After the first test, you should only have 1 Contact where the Nickname is “Jess”. Create a 2nd Contact (First Name = Test; Last Name = Contact; Nickname = Jessica).
2. Go to the Contact Merge Tab and click the Search Contacts button. Search for “Test Contact”. 
3. Select both records in the found set and hit Next.
4. The record selected within the Master Record row should be the 2nd one you Created since it defaults to the record with the most recent modified date. Its Nickname value of “Jessica” should also be selected.
5. Select the Master Record radio button for the first Contact (not the default “master” record) and also select its Nickname value (“Jess”). Do not hit the Select All option.
6. Click the Merge button and hit Merge in the warning modal.
7. Confirm the merged Contact has a value of “Jessica” instead of “Jess”.

***

*Test 4: (Works as expected): Keep the Nickname value from the default “Master” record.*

*Expected Behavior:* the merged Contact will have “JLo” as the Nickname value.
*Actual Behavior*: the merged Contact has “JLo” as the Nickname value.

1. After the first test, you should only have 1 Contact where the Nickname is “Jess”. Create a 2nd Contact (First Name = Test; Last Name = Contact; Nickname = JLo).
2. Go to the Contact Merge Tab and click the Search Contacts button. Search for “Test Contact”. 
3. Select both records in the found set and hit Next.
4. The record selected within the Master Record row should be the 2nd one you Created since it defaults to the record with the most recent modified date. Its Nickname value of “JLo” should also be selected.
5. Keep the values as is.
6. Click the Merge button and hit Merge in the warning modal.
7. Confirm the merged Contact has a value of “JLo” instead of “Jessica”.

**Comment by jesslopez-sf** [comment]
W-9313278

**Comment by karenzelevinsky** [comment]
Does this problem happen for all field types or only a subset?

**Comment by jesslopez-sf** [comment]
@karenzelevinsky As far as I know, it would affect all field types. However, if the field value on the "master" record was empty, our Contact Merge typically defaults to the other record that has a value when it loads. So it's possible that in those instances, the non-"master" value was saved to the "master" record.

**Comment by jesslopez-sf** [comment]
**Update**: this will be included in an emergency patch release:
Patch Release 3.198.1 Deployed to Sandboxes -- Wednesday, May 26
Patch Release 3.198.1 Deployed to Production -- Thursday, May 27

See Power of Us Hub post for more info:
https://powerofus.force.com/s/feed/0D57y00000FIZZoCAP

**Comment by jesslopez-sf** [comment]
**To remove access to the NPSP Contact Merge feature prior to our fix going out** -
The easiest way would be to remove Profile/Permission Set access to the Visualforce page called npsp.CON_ContactMerge. 
(i.e. go to Setup --> Profiles --> click on a Profile name --> click on the Visualforce page section --> remove npsp.CON_ContaactMerge). This will display an Insufficient Access error to the User and will prevent them from accessing the feature from buttons, tabs, etc.

**Comment by judisohn** [comment]
Resolved in patch release 3.198.1 through #6438 (pushing to production evening Thursday, 5/27)
